### PR TITLE
Keep retrying storing/deletng reminders

### PR DESCRIPTION
**Pull request by DeepanshuA**
# Description

When trying to run multiple Workflow instances (I used 25 WF instance per App Instance) in multiple App instances (used 2 App instance as of now), then even after change from https://github.com/dapr/dapr/pull/6885, workflows were getting 1. completely stuck AND 2. there were erroneous errors.

This PR on top of https://github.com/dapr/dapr/pull/6885 will ensure that Creation or Deletion of reminders happens definitely. And, thus resolves both issues:
1. Workflows were getting stuck, if Reminder was not created to schedule a workflow itself. Ensuring that reminder is created on retrying resolved this case.
2. The errors were coming as some reminders could not be deleted but the related workflow activities got completed. So, the errors like following were coming:
```
level=error msg="activity-processor: failed to process work item: invalid TaskScheduled event"
```
As deletion of reminder is ensured, this error is also resolved.

Confirmed on multiple runs that finally ALL workflow instances were getting Successfully Completed.

## Issue reference

Please reference the issue this PR (along with https://github.com/dapr/dapr/pull/6885) will close: #6874

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/actors/reminders/reminders.go [hunk 1]
```diff
@@ -159,9 +160,19 @@ func (r *reminders) CreateReminder(ctx context.Context, reminder *internal.Remin
 
 	stop := make(chan struct{})
 
-	err = r.storeReminder(ctx, store, reminder, stop)
+	config := retry.DefaultConfig()
```
Comment (ItalyPaleAle) [inline pkg/actors/reminders/reminders.go:163]: Does this have a maximum attempt count?

File: pkg/actors/reminders/reminders.go [hunk 2]
```diff
@@ -159,9 +160,19 @@ func (r *reminders) CreateReminder(ctx context.Context, reminder *internal.Remin
...
+	config.Multiplier = 1.0
+	b := config.NewBackOffWithContext(ctx)
+
+	err = retry.NotifyRecover(func() error {
+		return r.storeReminder(ctx, store, reminder, stop)
```
Comment (ItalyPaleAle) [inline pkg/actors/reminders/reminders.go:168]: I wonder if we should retry only if the error is related to etags? We have a way to see if the error is an etag error (which is implemented in components-contrib and has conformance tests): https://vscode.dev/github/ItalyPaleAle/dapr/blob/master/pkg/grpc/api.go#L703-L707

File: pkg/actors/reminders/reminders.go (hunk 1 above)
Comment (DeepanshuA) [reply pkg/actors/reminders/reminders.go:163]: no, I skipped it intentionally to ensure that we definitely store/delete reminder

File: pkg/actors/reminders/reminders.go (hunk 2 above)
Comment (DeepanshuA) [reply pkg/actors/reminders/reminders.go:168]: makes sense - let me try that

File: pkg/actors/reminders/reminders.go (hunk 1 above)
Comment (DeepanshuA) [reply pkg/actors/reminders/reminders.go:163]: And, I kept it Multiplier as 1.0 intentionally, to ensure that workflows execution time doesn't exceed too much.

File: pkg/actors/reminders/reminders.go (hunk 2 above)
Comment (DeepanshuA) [reply pkg/actors/reminders/reminders.go]: done

File: pkg/actors/reminders/reminders.go [hunk 3]
```diff
@@ -159,9 +160,37 @@ func (r *reminders) CreateReminder(ctx context.Context, reminder *internal.Remin
...
+		innerErr := r.storeReminder(ctx, store, reminder, stop)
+		if innerErr != nil {
+			if errors.As(innerErr, &etagErr) {
+				switch etagErr.Kind() {
+				case state.ETagMismatch:
+					// If the etag is mismatched, we can retry the operation.
+					return innerErr
+				default:
+					log.Errorf("error storing reminder: %w", innerErr)
+					err = innerErr
```
Comment (ItalyPaleAle) [inline pkg/actors/reminders/reminders.go:178]: Please don't do this: it's a side effect that will cause race conditions.

Instead, please use `NotifyRecoverWIthData` to return a second error if needed.

File: pkg/actors/reminders/reminders.go [hunk 4]
```diff
@@ -205,7 +234,40 @@ func (r *reminders) DeleteReminder(ctx context.Context, req internal.DeleteRemin
...
+					// we can't retry the operation.
+					return nil
+				}
+			}
+		}
+		log.Errorf("error storing reminder: %w", err)
+		err = innerErr
```
Comment (ItalyPaleAle) [inline pkg/actors/reminders/reminders.go:259]: Here too. Please, no side effects in goroutines. Use `NotifyRecoverWIthData`

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: pkg/actors/reminders/reminders.go (hunk 3 above)
Comment (ItalyPaleAle) [reply pkg/actors/reminders/reminders.go:178]: Actually even better... wrap the non-etag error in a permanent error: https://pkg.go.dev/github.com/cenkalti/backoff/v4#PermanentError we do that in other parts of Dapr.

**Comment by mukundansundar** [comment]
created a separate issue for tracking E2E for this change https://github.com/dapr/dapr/issues/6889

File: pkg/actors/reminders/reminders.go (hunk 3 above)
Comment (DeepanshuA) [reply pkg/actors/reminders/reminders.go]: done

File: pkg/actors/reminders/reminders.go (hunk 4 above)
Comment (DeepanshuA) [reply pkg/actors/reminders/reminders.go]: done

File: pkg/actors/reminders/reminders.go [hunk 5]
```diff
@@ -159,9 +161,33 @@ func (r *reminders) CreateReminder(ctx context.Context, reminder *internal.Remin
 
 	stop := make(chan struct{})
 
-	err = r.storeReminder(ctx, store, reminder, stop)
+	config := retry.DefaultConfig()
+	config.Multiplier = 1.0
+	b := config.NewBackOffWithContext(ctx)
+
+	var etagErr *state.ETagError
```
Comment (ItalyPaleAle) [inline pkg/actors/reminders/reminders.go:168]: This is a side effect too

File: pkg/actors/reminders/reminders.go [hunk 6]
```diff
@@ -159,9 +161,33 @@ func (r *reminders) CreateReminder(ctx context.Context, reminder *internal.Remin
...
+			if errors.As(innerErr, &etagErr) {
+				switch etagErr.Kind() {
+				case state.ETagMismatch:
+					// If the etag is mismatched, we can retry the operation.
+					return innerErr
+				default:
+					log.Errorf("error storing reminder: %w", innerErr)
+					return backoff.Permanent(innerErr)
+				}
+			}
```
Comment (ItalyPaleAle) [inline pkg/actors/reminders/reminders.go:181]: Missing a`return innerErr` here

**Review by ItalyPaleAle — APPROVED** [review]

**Review by artursouza — APPROVED** [review]

**Comment by ItalyPaleAle** [comment]
@DeepanshuA we should consider changing the log from warning to info. We are seeing on Discord users that tried the nightly and are concerned with this. Maybe even mask the error message and just say that we'll retry?

**Pull request merged**